### PR TITLE
feat: sequence-based ibc transfer detection

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

See this TODO for context:

https://github.com/shapeshift/web/blob/22e8548fa57f4a0bb59a6c70949c379993fc494b/src/lib/swapper/swappers/OsmosisSwapper/OsmosisSwapper.ts#L312-L313

Leverages the sequence number on both the initiating and destination chain to achieve *reliable* polling for Osmosis swaps.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes https://github.com/shapeshift/web/issues/4962

## Risk

Osmosis swaps execution / status may be borked, which is a non-concern as Osmosis is currently off

## Testing

- Osmosis swaps are happy on both sides, and trade status is reliable

### Engineering

- ☝🏽 
- Test with https://github.com/shapeshift/web/pull/5006 which turns Osmosis swapper back on
- Note, Osmosis node is borked on dev so make sure to `yarn env app` while testing this

### Operations

- ☝🏽 

## Screenshots (if applicable)

Screenshot 2023-07-31 at 22 59 05
Screenshot 2023-07-31 at 22 42 06

### CODE REVIEW

**Comment by gomesalexandre** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #4922** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4922" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/4922?utm_source=stack-comment).

File: src/lib/swapper/swappers/OsmosisSwapper/utils/helpers.ts [hunk 1]
```diff
@@ -35,26 +36,45 @@ type FindPoolOutput = {
   buyAssetIndex: number
 }
 
-const txStatus = async (txid: string, baseUrl: string): Promise<string> => {
+const fetchLcdTx = async (txid: string, baseUrl: string): Promise<any> => {
   try {
+    debugger
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/OsmosisSwapper/utils/helpers.ts:41]: Will go away, keeping for dev for now

File: src/lib/swapper/swappers/OsmosisSwapper/utils/helpers.ts [hunk 2]
```diff
@@ -71,6 +91,79 @@ export const pollForComplete = (txid: string, baseUrl: string): Promise<string>
...
+  baseUrl: string
+}): Promise<string> => {
+  return new Promise((resolve, reject) => {
+    const timeout = 300000 // 5 mins
+    const startTime = Date.now()
+    const interval = 5000 // 5 seconds
+
+    const poll = async function () {
+      const initiatingChainStatus = await fetchLcdTxStatus(txid, baseUrl)
+      const tx = await fetchLcdTx(txid, baseUrl)
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/OsmosisSwapper/utils/helpers.ts:113]: Tx could be injected to avoid double-fetching

File: src/lib/swapper/swappers/OsmosisSwapper/OsmosisSwapper.ts [hunk 3]
```diff
@@ -422,6 +426,21 @@ export class OsmosisSwapper implements Swapper<ChainId> {
         gas: ibcFromOsmosisFeeData.fast.chainSpecific.gasLimit,
         feeDenom: 'uosmo',
       })
+
+      // wait till confirmed
+      const pollResult = await pollForCrossChainComplete({
+        txid: tradeId,
+        baseUrl: osmoUrl,
+      })
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/OsmosisSwapper/OsmosisSwapper.ts:434]: We were never actually polling for OSMO-side Tx for ATOM -> OSMO but rather being optimistic, which is obviously no bueno

File: src/lib/swapper/swappers/OsmosisSwapper/utils/helpers.ts [hunk 4]
```diff
@@ -71,6 +72,73 @@ export const pollForComplete = (txid: string, baseUrl: string): Promise<string>
   })
 }
 
+// More logic here, since an IBC transfer consists of 2 transactions
+// 1. MsgTransfer on the initiating chain i.e send from source chain to destination chain
+// 2. MsgRecvPacket on the receiving chain i.e receive on destination chain from source chain
+// While 1. can simply be polled for (which we do on the method above),
+// the destination Tx needs to be picked by validators on the destination chain, and we don't know anything about said Tx in advance
+// TODO(gomes): Now that we're relying on Txhistory Txs, we could make this a usePoll hook, reactive on the TxHistory slice?
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/OsmosisSwapper/utils/helpers.ts:80]: Probably not going to happen, but leaving the TODO here just in case

File: src/state/slices/txHistorySlice/utils.ts [hunk 5]
```diff
@@ -47,3 +53,24 @@ export const serializeTxIndex = (
 
   return [accountId, txid, address.toLowerCase()].join(UNIQUE_TX_ID_DELIMITER)
 }
+
+export const deserializeTxIndex = (txIndex: TxIndex): TxDescriptor => {
```
Comment (gomesalexandre) [inline src/state/slices/txHistorySlice/utils.ts:57]: Unused in the end, but this guy will definitely come handy so leaving it here unless we decided it's scope creep

**Comment by 0xdef1cafe** [comment]
draft until osmo is healthy

File: src/components/MultiHopTrade/hooks/useTradeExecution/useTradeExecution.ts [hunk 6]
```diff
@@ -153,6 +153,10 @@ export const useTradeExecution = ({
           quoteId: tradeQuote.id,
           txHash: sellTxHashRef.current,
           chainId,
+          stepIndex: activeStepOrDefault,
+          quoteSellAssetAccountId: sellAssetAccountId,
+          quoteBuyAssetAccountId: buyAssetAccountId,
+          getState: store.getState,
```
Comment (gomesalexandre) [inline src/components/MultiHopTrade/hooks/useTradeExecution/useTradeExecution.ts:159]: @woodenfurniture FYI this is effectively becoming `checkStepStatus` - we might want to revamp the vernacular here

File: src/lib/swapper/swappers/OsmosisSwapper/utils/poll.ts [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+// the destination Tx needs to be picked by validators on the destination chain, and we don't know anything about said Tx in advance
+// TODO(gomes): Now that we're relying on Txhistory Txs, we could make this a usePoll hook, reactive on the TxHistory slice?
+export const pollForCrossChainComplete = ({
+  initiatingChainTxid,
+  initiatingChainAccountId,
+  getState,
+}: {
+  initiatingChainAccountId: AccountId
+  initiatingChainTxid: string
+  // Injecting this since we can't avoid circular dependencies when calling this from OsmosisSwapper/endpoints
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/OsmosisSwapper/utils/poll.ts:53]: Pretty much what the comment says

File: src/lib/swapper/swappers/OsmosisSwapper/utils/poll.ts
Comment (gomesalexandre) [inline src/lib/swapper/swappers/OsmosisSwapper/utils/poll.ts:1]: Moved to its own module because of circular deps

File: src/lib/swapper/swappers/OsmosisSwapper/endpoints.ts [hunk 8]
```diff
@@ -155,29 +160,65 @@ export const osmosisApi: Swapper2Api = {
...
+          stepSellAssetAccountId,
+          txHash,
+          fromAccountId(stepSellAssetAccountId).account,
+        )
+        const pollResult = await pollForCrossChainComplete({
+          initiatingChainTxid,
+          initiatingChainAccountId: stepSellAssetAccountId,
+          getState,
+        })
+        const status = pollResult === 'success' ? TxStatus.Confirmed : TxStatus.Failed
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/OsmosisSwapper/endpoints.ts:192]: I've had success on the ATOM -> OSMO first hop step status here, unfortunately the node gave up before I had a chance to test the second branch. Since Osmosis flag is off at the moment, I'd say we get this in as-is for now and test it once nodes are happy before turning the flag back on (or maybe keep it off? 👋🏽 )

File: src/lib/swapper/swappers/OsmosisSwapper/OsmosisSwapper.ts [hunk 9]
```diff
@@ -297,10 +299,17 @@ export class OsmosisSwapper implements Swapper<ChainId> {
...
+
       cosmosIbcTradeId = tradeId
 
       // wait till confirmed
-      const pollResult = await pollForComplete(tradeId, cosmosUrl)
+      const pollResult = await pollForCrossChainComplete({
+        initiatingChainTxid,
+        initiatingChainAccountId,
+        getState: store.getState,
+      })
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/OsmosisSwapper/OsmosisSwapper.ts:312]: Implementation was originally brought to single-hop swapper, kept it here so that both use the correct implementation, but this is effectively dead code now

File: src/lib/swapper/swappers/OsmosisSwapper/endpoints.ts (hunk 8 above)
Comment (0xApotheosis) [reply src/lib/swapper/swappers/OsmosisSwapper/endpoints.ts:192]: I was able to get both directions working successfully on the production node this evening 👌

**Review by 0xApotheosis — APPROVED** [review]
Looking good, Osmosis swapper is great again 🤝 

Tested Osmo to Atom, and Atom to Osmo. Transaction history/parsing and status updates during trade all worked correctly, and the expected amount was received in the expected currency.

Screenshot 2023-08-01 at 8 00 19 pm

Screenshot 2023-08-01 at 8 01 10 pm
